airflow/jobs.py

```diff
@@ -1122,9 +1124,26 @@ def _execute(self):
         )
         self.process = subprocess.Popen(['bash', '-c', command])
         return_code = None
+        last_heartbeat = datetime.now()
+        print("running faster version")
         while return_code is None:
-            self.heartbeat()
+            time_since_last_heartbeat = (datetime.now() -
+                                         last_heartbeat).total_seconds()
+            time_until_next_heartbeat = \
+                int(max(0, self.heartrate - time_since_last_heartbeat))
+
+            # Wait for task to finish running, up until we need to heartbeat.
+            try:
+                with timeout(time_until_next_heartbeat):
+                    self.process.wait()
+            except AirflowTaskTimeout:
+                pass
+
             return_code = self.process.poll()
+            # If the task finished, there's no need to heartbeat.
+            if return_code is not None:
```

Comment:
Nit: A little bit cleaner if the heartbeat is placed into the if and the "not" and "break" are removed (since the loop will exit right after due to the same condition)